nltk/test/unit/test_util.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+"""
+Unit tests for nltk.util.
+"""
+
+import unittest
+from nltk.util import everygrams
+
+
+class TestEverygrams(unittest.TestCase):
+
+    def setUp(self):
+        """Form test data for tests."""
+        self.test_data = iter('a b c'.split())
+
+    def test_everygrams_without_padding(self):
+        expected_output = [
+            ('a',),
+            ('a', 'b'),
+            ('a', 'b', 'c'),
+            ('b',),
+            ('b', 'c'),
+            ('c',),
+        ]
+        output = everygrams(self.test_data)
+        self.assertCountEqual(output, expected_output)
+
+    def test_everygrams_max_len(self):
+        expected_output = [('a',), ('a', 'b'), ('b',), ('b', 'c'),('c',),]
+        output = everygrams(self.test_data, max_len=2)
+        self.assertCountEqual(output, expected_output)
+
+    def test_everygrams_min_len(self):
+        expected_output = [('a', 'b'), ('b', 'c'), ('a', 'b', 'c'),]
+        output = everygrams(self.test_data, min_len=2)
+        self.assertCountEqual(list(output), expected_output)
+
+    def test_everygrams_pad_right(self):
+        expected_output = [
+            ('a',),
+            ('a', 'b'),
+            ('a', 'b', 'c'),
+            ('b',),
+            ('b', 'c'),
+            ('b', 'c', None),
+            ('c',),
+            ('c', None),
+            ('c', None, None),
+            (None,),
+            (None, None),
+            (None,),
+        ]
+        output = everygrams(self.test_data, max_len=3, pad_right=True)
+        self.assertCountEqual(list(output), expected_output)
+
+    def test_everygrams_pad_left(self):
+        expected_output = [
+            (None,),
+            (None, None),
+            (None, None, 'a'),
+            (None,),
```

Comment:
Ok, one last sanity-check: we are not surprised that this `(None,)` item appears for the second time, right? This is, if I understand correctly, the major difference with `padded_everygrams` from the lm module.

Reply:
The same happens for `padded_everygrams` from the the lm module. For example, the following test passes:

```python
def test_padded_everygrams(self):
        expected_output = [
            ('<s>',),
            ('<s>', '<s>'),
            ('<s>', '<s>', 'a'),
            ('<s>',),
            ('<s>', 'a'),
            ('<s>', 'a', 'b'),
            ('a',),
            ('a', 'b'),
            ('a', 'b', 'c'),
            ('b',),
            ('b', 'c'),
            ('b', 'c', '</s>'),
            ('c',),
            ('c', '</s>'),
            ('c', '</s>', '</s>'),
            ('</s>',),
            ('</s>', '</s>'),
            ('</s>',),
        ]
        test_data = 'a b c'.split()
        output = padded_everygrams(3, iter(test_data))
        self.assertCountEqual(output, expected_output)
        output = everygrams(iter(test_data), max_len=3, pad_left=True, pad_right=True, left_pad_symbol='<s>', right_pad_symbol='</s>')
        self.assertCountEqual(output, expected_output)
```
`padded_everygrams` performs `everygrams` with specific defaults.

Reply:
I see. Very clear, thank you!